pkg/registry/client.go

```diff
@@ -209,74 +186,154 @@ func ClientOptPlainHTTP() ClientOption {
 	}
 }
 
-// ClientOptResolver returns a function that sets the resolver setting on a client options set
-func ClientOptResolver(resolver remotes.Resolver) ClientOption {
-	return func(client *Client) {
-		client.resolver = func(ref registry.Reference) (remotes.Resolver, error) {
-			return resolver, nil
-		}
-	}
-}
-
 type (
 	// LoginOption allows specifying various settings on login
-	LoginOption func(*loginOperation)
-
-	loginOperation struct {
-		username string
-		password string
-		insecure bool
-		certFile string
-		keyFile  string
-		caFile   string
-	}
+	LoginOption func(host string, client *Client) error
 )
 
 // Login logs into a registry
 func (c *Client) Login(host string, options ...LoginOption) error {
-	operation := &loginOperation{}
 	for _, option := range options {
-		option(operation)
+		if err := option(host, c); err != nil {
+			return fmt.Errorf("configuring login option: %w", err)
+		}
+	}
+
+	loginURL := url.URL{
+		Scheme: "https",
+		Host:   host,
+		Path:   "v2",
+	}
+	if c.plainHTTP {
+		loginURL.Scheme = "http"
+	}
+
+	req, err := http.NewRequest(http.MethodGet, loginURL.String(), nil)
+	if err != nil {
+		return fmt.Errorf("creating request to %q: %w", &loginURL, err)
+	}
+
+	res, err := c.authorizer.Do(req)
+	if err != nil {
+		return fmt.Errorf("sending request to %q: %w", &loginURL, err)
```

Comment:
How about:
```suggestion
	reg, err := remote.NewRegistry(host)
	if err != nil {
		// handle error
	}
	reg.Client = c.authorizer
	ctx := context.Background()
	cred, err := c.authorizer.Credential(ctx, host)
	if err != nil {
		// handle error
	}
	if err := credentials.Login(ctx, c.store, reg, cred); err != nil {
		// handle error
	}
	fmt.Fprintln(c.out, "Login Succeeded")
```

Reply:
Opted to use [`Registry.Ping`](https://pkg.go.dev/oras.land/oras-go/v2@v2.2.1/registry/remote#Registry.Ping)

Reply:
One tricky thing is that docker CLI stores credentials of `docker.io` under the key `https://index.docker.io/v1/`. 
`credentials.Login` automatically handles this mapping but `Store.Put` does not (see https://github.com/oras-project/oras-credentials-go/blob/v0.3.0/registry.go#L55). 

You may consider using [credentials.ServerAddressFromRegistry](https://pkg.go.dev/github.com/oras-project/oras-credentials-go#ServerAddressFromRegistry) before putting the credential.